linera-base/src/crypto/mod.rs

```diff
@@ -46,87 +78,183 @@ pub type ValidatorKeypair = secp256k1::Secp256k1KeyPair;
+        match self {
+            AccountPublicKey::Ed25519(_) => SignatureScheme::Ed25519,
+            AccountPublicKey::Secp256k1(_) => SignatureScheme::Secp256k1,
+        }
     }
 
     /// Returns the byte representation of the public key.
-    pub fn to_bytes(&self) -> Vec<u8> {
-        self.0.as_bytes()
+    pub fn as_bytes(&self) -> Vec<u8> {
+        let mut bytes: Vec<u8> = Vec::new();
+        bytes.push(self.scheme().flag());
+
+        match self {
+            AccountPublicKey::Ed25519(public_key) => {
+                bytes.extend_from_slice(&public_key.as_bytes())
+            }
+            AccountPublicKey::Secp256k1(public_key) => {
+                bytes.extend_from_slice(&public_key.as_bytes())
+            }
+        }
+
+        bytes
     }
 
     /// Parses the byte representation of the public key.
     ///
-    /// Returns error if the byte slice has incorrect length.
+    /// Returns error if the byte slice has incorrect length or the flag is not recognized.
     pub fn from_slice(bytes: &[u8]) -> Result<Self, CryptoError> {
-        Ok(Self(ed25519::Ed25519PublicKey::from_slice(bytes)?))
+        let flag = bytes[0];
+        let scheme = SignatureScheme::from_flag_byte(flag)?;
+        match scheme {
+            SignatureScheme::Ed25519 => {
+                let pk = Ed25519PublicKey::from_slice(&bytes[1..])?;
+                Ok(AccountPublicKey::Ed25519(pk))
+            }
+            SignatureScheme::Secp256k1 => {
+                let pk = secp256k1::Secp256k1PublicKey::from_bytes(&bytes[1..])?;
+                Ok(AccountPublicKey::Secp256k1(pk))
+            }
+        }
     }
-}
 
-impl AccountSignature {
-    /// Creates a signature for the `value` using provided `secret`.
-    pub fn new<'de, T>(value: &T, secret: &AccountSecretKey) -> Self
-    where
-        T: BcsSignable<'de>,
-    {
-        let signature = ed25519::Ed25519Signature::new(value, &secret.0);
-        AccountSignature(signature)
+    /// A fake public key used for testing.
+    #[cfg(with_testing)]
+    pub fn test_key(name: u8) -> Self {
+        AccountPublicKey::Ed25519(Ed25519PublicKey::test_key(name))
     }
+}
 
+impl AccountSignature {
     /// Verifies the signature for the `value` using the provided `public_key`.
     pub fn verify<'de, T>(&self, value: &T, author: AccountPublicKey) -> Result<(), CryptoError>
     where
         T: BcsSignable<'de> + std::fmt::Debug,
     {
-        self.0.check(value, author.0)
+        match self {
+            AccountSignature::Ed25519(signature) => {
+                let public_key = match author {
+                    AccountPublicKey::Ed25519(public_key) => public_key,
+                    _ => {
+                        return Err(CryptoError::InvalidSignature {
+                            error: "invalid public key type".to_string(),
+                            type_name: std::any::type_name::<T>().to_string(),
+                        })
+                    }
```

Comment:
It seems like we could `match (self, author) { .. }` and avoid the code duplication